src/control/lib/cache/cache.go

```diff
@@ -0,0 +1,220 @@
+//
+// (C) Copyright 2023 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package cache
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/common"
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+type Item interface {
+	Lock()
+	Unlock()
+	Key() string
+	Refresh(ctx context.Context) error
+	NeedsRefresh() bool
+}
+
+// ItemCache is a mechanism for caching Items to keys.
+type ItemCache struct {
+	log   logging.Logger
+	mutex sync.RWMutex
+	items map[string]Item
+}
+
+// NewItemCache creates a new ItemCache.
+func NewItemCache(log logging.Logger) *ItemCache {
+	c := &ItemCache{
+		log:   log,
+		items: make(map[string]Item),
+	}
+	return c
+}
+
+// Set caches an item under a given key.
+func (ic *ItemCache) Set(item Item) error {
+	if ic == nil {
+		return errors.New("ItemCache is nil")
+	}
+
+	if common.InterfaceIsNil(item) || item.Key() == "" {
+		return errors.New("invalid item")
+	}
+
+	ic.mutex.Lock()
+	defer ic.mutex.Unlock()
+
+	ic.set(item)
+	return nil
+}
+
+func (ic *ItemCache) set(item Item) {
+	ic.items[item.Key()] = item
+}
+
+// Delete fully deletes an Item from the cache.
+func (ic *ItemCache) Delete(key string) {
+	if ic == nil {
+		return
+	}
+
+	ic.mutex.Lock()
+	defer ic.mutex.Unlock()
+
+	delete(ic.items, key)
+}
+
+// Has checks whether any item is cached under the given key.
+func (ic *ItemCache) Has(key string) bool {
+	if ic == nil {
+		return false
+	}
+
+	ic.mutex.RLock()
+	defer ic.mutex.RUnlock()
+
+	_, found := ic.items[key]
+	return found
+}
+
+// Keys returns a sorted list of all keys in the cache.
+func (ic *ItemCache) Keys() []string {
+	if ic == nil {
+		return nil
+	}
+
+	keys := []string{}
+	for k := range ic.items {
```

Comment:
From what I know range is not thread safe and thus a lock (or using a channel to access keys) should be needed ?

Reply:
+1, looks like it needs RLock like `Has()` method above.

Reply:
Good catch.